drivers/can/stm32_ll_can.c

```diff
@@ -0,0 +1,752 @@
+		k_sem_give(&data->tx_int_sem);
+	}
+	while (can->RF0R & CAN_RF0R_FMP0_Msk) {
+		struct can_msg msg;
+		CAN_FIFOMailBox_TypeDef *mbox = &can->sFIFOMailBox[0];
+		int filter_match_index = ((mbox->RDTR & CAN_RDT0R_FMI_Msk)
+					  >> CAN_RDT0R_FMI_Pos);
+
+		if (filter_match_index >= CONFIG_CAN_MAX_FILTER) {
+			break;
+		}
+
+		SYS_LOG_DBG("Message on filter index %d", filter_match_index);
+		can_get_msg_fifo(mbox, &msg);
+		if (data->rx_response[filter_match_index]) {
+			if (data->response_type & (1ULL << filter_match_index)) {
+				struct k_msgq *msg_q = data->rx_response[filter_match_index];
+
+				k_msgq_put(msg_q, &msg, K_NO_WAIT);
+			} else {
+				can_rx_callback_t callback =
+					data->rx_response[filter_match_index];
+				callback(&msg);
+			}
+		}
+
+		/* Release message */
+		can->RF0R |= CAN_RF0R_RFOM0;
+	}
+}
+
+void HAL_CAN_MspInit(CAN_HandleTypeDef *hcan)
+{
+	ARG_UNUSED(hcan);
+}
+static int can_stm32_init(struct device *dev)
+{
+	const struct can_stm32_config *cfg = DEV_CFG(dev);
+	struct can_stm32_data *data = DEV_DATA(dev);
+	CAN_TypeDef *can = cfg->can;
+	struct device *clock;
+	int ret;
+
+	k_mutex_init(&data->tx_mutex);
+	k_mutex_init(&data->set_filter_mutex);
+	k_sem_init(&data->tx_int_sem, 0, 1);
+	k_sem_init(&data->mb0.tx_int_sem, 0, 1);
+	k_sem_init(&data->mb1.tx_int_sem, 0, 1);
+	k_sem_init(&data->mb2.tx_int_sem, 0, 1);
+	data->mb0.tx_callback = NULL;
+	data->mb1.tx_callback = NULL;
+	data->mb2.tx_callback = NULL;
+
+	data->filter_usage = (1ULL << CAN_MAX_NUMBER_OF_FILTES) - 1ULL;
+	memset(data->rx_response, 0, sizeof(void *) * CONFIG_CAN_MAX_FILTER);
+	data->response_type = 0;
+
+	clock = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	__ASSERT_NO_MSG(clock);
+
+	ret = clock_control_on(clock, (clock_control_subsys_t *) &cfg->pclken);
+	if (ret != 0) {
+		SYS_LOG_ERR("HAL_CAN_Init clock control on failed: %d", ret);
+		return -EIO;
+	}
+
+	ret = can_stm32_runtime_configure(dev, CAN_NORMAL_MODE, 0);
+	if (ret) {
+		return ret;
+	}
+
+	cfg->config_irq(can);
+	can->IER |= CAN_IT_TME;
+	SYS_LOG_INF("Init of %s done", dev->config->name);
+	return 0;
+}
+
+int can_stm32_runtime_configure(struct device *dev, can_mode_t mode,
+				u32_t bitrate)
+{
+	CAN_HandleTypeDef hcan;
+	const struct can_stm32_config *cfg = DEV_CFG(dev);
+	CAN_TypeDef *can = cfg->can;
+	struct device *clock;
+	u32_t clock_rate;
+	u32_t prescaler;
+	u32_t hal_mode;
+	int hal_ret;
+
+	clock = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	__ASSERT_NO_MSG(clock);
+	hcan.Instance = can;
+
+	clock_control_get_rate(clock, (clock_control_subsys_t *) &cfg->pclken,
+			       &clock_rate);
+	if (!bitrate) {
+		bitrate = cfg->bus_speed;
+	}
+	prescaler = clock_rate / (BIT_SEG_LENGTH * cfg->bus_speed);
+	if (prescaler == 0 || prescaler > 1024) {
+		SYS_LOG_ERR("HAL_CAN_Init failed: prescaler > max (%d > 1024)",
+					prescaler);
+		return -EINVAL;
+	}
+
+	const u32_t bs1 = (CONFIG_CAN_PHASE_SEG1_PROP_SEG - 1) << CAN_BTR_TS1_Pos;
```

Comment:
Sometimes it is tempting thing to do, but in zephyr we try to avoid declaring variables in the middle of the function